presto-geospatial/src/main/java/com/facebook/presto/geospatial/GeoFunctions.java

```diff
@@ -0,0 +1,585 @@
+        return leftGeometry.within(rightGeometry);
+    }
+
+    private static OGCGeometry getGeometry(OGCGeometry geometry, String function, String...validTypes)
+    {
+        requireNonNull(geometry, "geometry is null");
+        String type = geometry.geometryType();
+        for (String validType : validTypes) {
+            if (type.equals(validType)) {
+                return geometry;
+            }
+        }
+        throw new IllegalArgumentException(function + " only applies to " + StringUtils.join(asList(validTypes), " or ") + ". Input type is: " + type);
+    }
+
+    private static Envelope getEnvelope(OGCGeometry geometry)
+    {
+        Envelope envelope = new Envelope();
+        geometry.getEsriGeometry().queryEnvelope(envelope);
+        return envelope;
+    }
+
+    private static void verifySameSpatialReference(OGCGeometry leftGeometry, OGCGeometry rightGeometry)
+    {
+        checkArgument(leftGeometry.getEsriSpatialReference().equals(rightGeometry.getEsriSpatialReference()), "Input Geometries Spatial Reference do not match");
+    }
+
+    // Points centroid is arithmetic mean of the input points
+    private static Point getPointsCentroid(OGCGeometry geometry)
+    {
+        MultiVertexGeometry multiVertex = (MultiVertexGeometry) geometry.getEsriGeometry();
+        double xSum = 0;
+        double ySum = 0;
+        for (int i = 0; i < multiVertex.getPointCount(); i++) {
+            Point point = multiVertex.getPoint(i);
+            xSum += point.getX();
+            ySum += point.getY();
+        }
+        return new Point(xSum / multiVertex.getPointCount(), ySum / multiVertex.getPointCount());
+    }
+
+    // Lines centroid is weighted mean of each line segment, weight in terms of line length
+    private static Point getLineCentroid(OGCGeometry geometry)
+    {
+        Polyline polyline = (Polyline) geometry.getEsriGeometry();
+        double xSum = 0;
+        double ySum = 0;
+        double weightSum = 0;
+        for (int i = 0; i < polyline.getPathCount(); i++) {
+            Point startPoint = polyline.getPoint(polyline.getPathStart(i));
+            Point endPoint = polyline.getPoint(polyline.getPathEnd(i) - 1);
+            Line line = new Line();
+            line.setStart(startPoint);
+            line.setEnd(endPoint);
+            double weight = line.calculateLength2D();
+            weightSum += weight;
+            xSum += (startPoint.getX() + endPoint.getX()) * weight / 2;
+            ySum += (startPoint.getY() + endPoint.getY()) * weight / 2;
+        }
+        return new Point(xSum / weightSum, ySum / weightSum);
+    }
+
+    // Polygon centroid:
+    // c[x] = (Sigma(x[i] + x[i + 1]) * (x[i] * y[i + 1] - x[i + 1] * y[i]), for i = 0 to N - 1) / (6 * signedArea)
+    // c[y] = (Sigma(y[i] + y[i + 1]) * (x[i] * y[i + 1] - x[i + 1] * y[i]), for i = 0 to N - 1) / (6 * signedArea)
+    private static Point getPolygonCentroid(OGCGeometry geometry)
```

Comment:
@zhenxiao , I just realized that this formula applies only polygons without holes. The general case of polygon with one or more holes requires a bit more computation. I haven't found an official formula yet, but the following Stack Overflow answer may be correct. They compute the centroid as a weighted diff between outer inner polygons.

https://math.stackexchange.com/questions/623841/finding-centroid-of-polygon-with-holes-polygons/623849#623849

Reply:
Seems that is more computation for it. How about we add a comment, and explicitly not support polygons with holes?